Add support for reading from sqlite calibration databases

hgreenlee opened the pull request.

hgreenlee committed:
Add ability to read calibration data from sqlite.

hgreenlee committed:
Remove active IOV constraint.

hgreenlee committed:
Merge branch 'feature/greenlee_sqlite_develop' of github.com:uboone/larevt into develop

hgreenlee committed:
Merge remote-tracking branch 'upstream/develop' into develop

hgreenlee committed:
Merge remote-tracking branch 'origin/develop' into feature/greenlee_sqlite_develop

hgreenlee committed:
Merge remote-tracking branch 'origin/develop' into feature/greenlee_sqlite_develop

hgreenlee committed:
Update DBFolder.

hgreenlee committed:
Change data type from string to boost::variant.

hgreenlee committed:
Switch from boost::variant to std::variant.

FNALbuild commented:
A new Pull Request was created by  @hgreenlee (Herbert Greenlee) for develop.

It involves the following packages:

larevt


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
A new Pull Request was created by  @hgreenlee (Herbert Greenlee) for develop.

It involves the following packages:

larevt


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//770/console

knoepfel commented on larevt/CalibrationDBI/Providers/DBDataset.cxx:
```diff
@@ -0,0 +1,254 @@
+// Destructor.
+
+lariov::DBDataset::~DBDataset()
+{}
```
Suggest removing the d'tor as it is not cleaning up anything.

knoepfel commented on larevt/CalibrationDBI/Providers/DBDataset.h:
```diff
@@ -0,0 +1,163 @@
+
+      // Accessors.
+
+      bool isValid() const {return fData != 0;}
```
Suggest using `fData != nullptr` to be more explicit.

knoepfel commented on larevt/CalibrationDBI/Providers/DBDataset.h:
```diff
@@ -0,0 +1,163 @@
+    std::vector<std::string>& colNames() {return fColNames;}
+    std::vector<std::string>& colTypes() {return fColTypes;}
+    std::vector<DBChannelID_t>& channels() {return fChannels;}
+    std::vector<value_type>& data() {return fData;}
```
Suggest removing accessors that allow modification of internal data.  Are they needed?

knoepfel commented on larevt/CalibrationDBI/Providers/DBDataset.cxx:
```diff
@@ -0,0 +1,254 @@
+	//std::cout << "DBDataset: row=" << row << ", column=" << col << ", value=" << fData.back() 
+	//	<< std::endl;
+	if(col == 0) {
+	  std::cout << "First column has wrong type real." << std::endl;
```
Suggest removing all `std::cout` statements.  If you absolutely require printout to the terminal, then the message-facility utilities should be used.  Otherwise, the exception message already includes the error message.

knoepfel commented on larevt/CalibrationDBI/Providers/DBDataset.h:
```diff
@@ -0,0 +1,163 @@
+
+    // Typedef
+
+    typedef std::variant<long, double, std::unique_ptr<std::string> > value_type;
```
The `std::unique_ptr<std::string>` type will always take more memory than simply `std::string` as it adds another layer of indirection through a pointer--it's an additional 8 bytes.  Switching to `std::variant<long, double, std::string>` has the effect of increasing the stack size of `std::variant<...>`, but in so doing it shifts what was heap memory (via the unique_ptr indirection) to stack memory.

knoepfel commented on larevt/CalibrationDBI/Providers/DBDataset.h:
```diff
@@ -0,0 +1,163 @@
+
+      // Modifiers.
+
+      void clear() {fData = 0; return;}
```
Simply `{ fData = nullptr; }` is sufficient.

knoepfel commented on larevt/CalibrationDBI/Providers/DBDataset.h:
```diff
@@ -0,0 +1,163 @@
+    void setBeginTime(const IOVTimeStamp& t) {fBeginTime = t;}
+    void setEndTime(const IOVTimeStamp& t) {fEndTime = t;}
+    void setNRows(size_t nrows) {fNRows = nrows;}
+    void setNCols(size_t ncols) {fNCols = ncols;}
```
Suggest removing as many modifiers as possible--for multithreading considerations.  If the `DBDataset` really needs to be cleared, a better approach is to do:
```.diff
- dbDataSet.clear()
+ dbDataSet = DBDataSet(...);
```

knoepfel commented on larevt/CalibrationDBI/Providers/DBFolder.cxx:
```diff
@@ -7,109 +7,156 @@
+
+  // Destructor.
+
+  DBFolder::~DBFolder() {}
```
Suggest removing unnecessary d'tor definition.

knoepfel commented on larevt/CalibrationDBI/Providers/DBFolder.cxx:
```diff
@@ -229,81 +255,639 @@ namespace lariov {
+    rc = sqlite3_prepare_v2(db, sql.str().c_str(), -1, &stmt, 0);
+    if(rc != SQLITE_OK) {
+      std::cout << "sqlite3_prepare_v2 failed." << fSQLitePath << std::endl;
+      std::cout << "Failed sql = " << sql.str() << std::endl;
```
Suggest removing `std::cout` statements just before exception throw--include the relevant message in the exception.

knoepfel commented on larevt/CalibrationDBI/Providers/DBFolder.cxx:
```diff
@@ -229,81 +255,639 @@ namespace lariov {
+  {
+    size_t nrows = data.nrows();
+    size_t ncols = data.ncols();
+    std::cout << "Dataset contains " << nrows << " rows and " << ncols << " columns." << std::endl;
```
Can you use messagefacility instead?

knoepfel commented on larevt/CalibrationDBI/Providers/DBFolder.cxx:
```diff
@@ -229,81 +255,639 @@ namespace lariov {
+    }
+  }
+
+  void DBFolder::CompareDataset(const DBDataset& data1, DBDataset& data2) const
```
Suggest returning `bool` so that manual inspection is not necessary--i.e. it would be ideal to type in a test:

```.cpp
assert(dbFolder.CompareDataset(db1, db2));
```
Also, `std::cout` statements should be converted to message-facility ones, and is there a reason the second function argument cannot be marked `const`?

knoepfel reviewed (commented):
Thanks for the PR, Herb.  A few comments included below for your consideration.

hgreenlee committed:
Code cleanups.

FNALbuild commented:
Pull request #9 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//774/console

lgarren commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

hgreenlee commented:
Updates from most recent commit to parent branch.

1.  Removed unnecessary destructors.
2.  Replaced all std::cout with message facility.
3.  Removed modify methods from DBDataset (now const after construction).
4.  Second argument of DBFolder::CompareDataset is now const.

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10106&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10105&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/5106&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10105&builds=

FNALbuild commented:
-argoneut tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/5307&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10105&builds=

FNALbuild commented:
-icarus tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/1349&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10105&builds=

FNALbuild commented:
-sbnd tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/5750&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10105&builds=

FNALbuild commented:
-dune tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/10548&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10105&builds=

FNALbuild commented:
-uboone tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5723&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/10105&builds=

lgarren commented:
appprove

This change was presented and approved at a coordination meeting.

lgarren commented:
approve

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests.
